### PR TITLE
Storage/STG78 OAuth Copy

### PR DESCRIPTION
**Pull request by microzchang**
# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/main/CONTRIBUTING.md#pull-requests).

- [ ] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [ ] Doxygen docs
- [ ] Unit tests
- [ ] No unwanted commits/changes
- [ ] Descriptive title/description
  - [ ] PR is single purpose
  - [ ] Related issue listed
- [ ] Comments in source
- [ ] No typos
- [ ] Update changelog
- [ ] Not work-in-progress
- [ ] External references or docs updated
- [ ] Self review of PR done
- [ ] Any breaking changes?

### CODE REVIEW

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp [hunk 1]
```diff
@@ -587,6 +587,11 @@ namespace Azure { namespace Storage { namespace Blobs {
      * in this option. Default is to replace.
      */
     Models::BlobCopySourceTagsMode CopySourceTagsMode;
+
+    /**
+     * @brief Optional. Source authentication used to access the source blob.
+     */
+    Azure::Nullable<std::string> SourceAuthentication;
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp:594]: according to this discussion, we should call this `SourceAuthorization`.

https://github.com/Azure/azure-sdk-for-cpp/issues/4718

File: sdk/storage/azure-storage-blobs/test/ut/append_blob_client_test.cpp [hunk 2]
```diff
@@ -349,4 +349,30 @@ namespace Azure { namespace Storage { namespace Test {
     EXPECT_EQ(ReadBodyStream(appendBlobClient.Download().Value.BodyStream), blockContent);
   }
 
+  TEST_F(AppendBlobClientTest, OAuthAppendBlockFromUri_LIVEONLY_)
```
Comment (Jinming-Hu) [inline sdk/storage/azure-storage-blobs/test/ut/append_blob_client_test.cpp:352]: I think we should update the sanitizer and make these test cases work in record-playback mode before merging this PR

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp [hunk 3]
```diff
@@ -587,6 +587,11 @@ namespace Azure { namespace Storage { namespace Blobs {
      * in this option. Default is to replace.
      */
     Models::BlobCopySourceTagsMode CopySourceTagsMode;
+
+    /**
+     * @brief Optional. Source authentication used to access the source blob.
+     */
+    Azure::Nullable<std::string> SourceAuthorization;
```
Comment (ahsonkhan) [inline sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp:594]: This doesn't need to be nullable. See https://github.com/Azure/azure-sdk-for-cpp/issues/4718#issuecomment-1641107074

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp [hunk 4]
```diff
@@ -587,6 +587,11 @@ namespace Azure { namespace Storage { namespace Blobs {
      * in this option. Default is to replace.
      */
     Models::BlobCopySourceTagsMode CopySourceTagsMode;
+
+    /**
+     * @brief Optional. Source authentication used to access the source blob.
```
Comment (ahsonkhan) [inline sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp:592]:
```suggestion
     * @brief Optional. Source authorization used to access the source blob.
```

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp [hunk 5]
```diff
@@ -953,6 +958,11 @@ namespace Azure { namespace Storage { namespace Blobs {
...
```
Comment (ahsonkhan) [inline sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp:963]:
```suggestion
     * @brief Optional. Source authorization used to access the source blob.
```

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp [hunk 6]
```diff
@@ -1001,6 +1011,11 @@ namespace Azure { namespace Storage { namespace Blobs {
     struct : public Azure::ModifiedConditions, public Azure::MatchConditions
     {
     } SourceAccessConditions;
+
+    /**
+     * @brief Optional. Source authentication used to access the source blob.
```
Comment (ahsonkhan) [inline sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp:1016]:
```suggestion
     * @brief Optional. Source authorization used to access the source blob.
```

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp [hunk 7]
```diff
@@ -1286,6 +1301,11 @@ namespace Azure { namespace Storage { namespace Blobs {
      * @brief Optional conditions that must be met to perform this operation.
      */
     AppendBlobAccessConditions AccessConditions;
+
+    /**
+     * @brief Optional. Source authentication used to access the source blob.
```
Comment (ahsonkhan) [inline sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp:1306]:
```suggestion
     * @brief Optional. Source authorization used to access the source blob.
```

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp [hunk 8]
```diff
@@ -1389,6 +1409,11 @@ namespace Azure { namespace Storage { namespace Blobs {
     struct : public Azure::ModifiedConditions, public Azure::MatchConditions
     {
     } SourceAccessConditions;
+
+    /**
+     * @brief Optional. Source authentication used to access the source blob.
```
Comment (ahsonkhan) [inline sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_options.hpp:1414]:
```suggestion
     * @brief Optional. Source authorization used to access the source blob.
```

**Review by Jinming-Hu — APPROVED** [review]

**Comment by microzchang** [comment]
pipeline: https://dev.azure.com/azure-sdk/internal/_build/results?buildId=2981287&view=results

**Pull request merged**